### Decision Tree :
- Simple Tree like structure, model makes a decision at every node
- Useful in simple tasks
- One of the most popular algorithm
- Easy explainability easy to show how a decision process works!
  

###  why decision Tree are popular :
- Easy to interpret and present
- Well defined logic mimic human level thought
- Random Forests, ensembles of decision trees are more powerful classifiers
- Feature values are preferred to be categorical . If the values are continuous then they are discretized prior to building the model
 


### Build decision Trees:
- Two common Algorithms:
- CART(Classification and Regression Trees) -->> Uses Gini Index (classification) as metric.
- ID3(iterative Dichotomiser 3 ) -->> uses Entropy function and information gain as metric

### Entropy: or (randomness)
- H(s) = - {summation (p_c)* log_2(p_c)}, where H is entropy , s is for system H(S) is entropy of system and  p_c is probability of class c.
- Entropy that can be maxitmum can be is 1 when everyclass  has the  same no. of examples in it.
- and Minimum can be 0. when there is only one class.
- Can randomness be negative?

H(S) = - {summation (p_c)*log_2 (p_c)}

### Data Pre-Processing 

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv("titanic/train.csv")

In [3]:
df.head(n =3)

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    male  22.0      1      0   7.2500        S
1         1       1  female  38.0      1      0  71.2833        C
2         1       3  female  26.0      0      0   7.9250        S

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 8 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    object 
 3   Age       714 non-null    float64
 4   SibSp     891 non-null    int64  
 5   Parch     891 non-null    int64  
 6   Fare      891 non-null    float64
 7   Embarked  889 non-null    object 
dtypes: float64(2), int64(4), object(2)
memory usage: 55.8+ KB


- #### what we will do is we will remove irrelevant data
- #### And remove NaN values with average age
- #### Age has only 714 components, It NaN values

In [5]:
df.columns

Index(['Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare',
       'Embarked'],
      dtype='object')

In [6]:
df = df.drop(["Ticket", "PassengerId", "Cabin","Name"], axis = 1)
# alreday dropped now

KeyError: "['Ticket', 'PassengerId', 'Cabin', 'Name'] not found in axis"

In [7]:
df

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0           0       3    male  22.0      1      0   7.2500        S
1           1       1  female  38.0      1      0  71.2833        C
2           1       3  female  26.0      0      0   7.9250        S
3           1       1  female  35.0      1      0  53.1000        S
4           0       3    male  35.0      0      0   8.0500        S
..        ...     ...     ...   ...    ...    ...      ...      ...
886         0       2    male  27.0      0      0  13.0000        S
887         1       1  female  19.0      0      0  30.0000        S
888         0       3  female   NaN      1      2  23.4500        S
889         1       1    male  26.0      0      0  30.0000        C
890         0       3    male  32.0      0      0   7.7500        Q

[891 rows x 8 columns]

In [8]:
avg_age = df["Age"].mean()
avg_age.round()

30.0

In [9]:
df['Age'].replace({"NaN": "30"})
df.to_csv("titanic/train.csv", index=False) 
df

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0           0       3    male  22.0      1      0   7.2500        S
1           1       1  female  38.0      1      0  71.2833        C
2           1       3  female  26.0      0      0   7.9250        S
3           1       1  female  35.0      1      0  53.1000        S
4           0       3    male  35.0      0      0   8.0500        S
..        ...     ...     ...   ...    ...    ...      ...      ...
886         0       2    male  27.0      0      0  13.0000        S
887         1       1  female  19.0      0      0  30.0000        S
888         0       3  female   NaN      1      2  23.4500        S
889         1       1    male  26.0      0      0  30.0000        C
890         0       3    male  32.0      0      0   7.7500        Q

[891 rows x 8 columns]

In [10]:
from sklearn.preprocessing import LabelEncoder

In [11]:
le = LabelEncoder()

In [12]:
df["Sex"] = le.fit_transform(df["Sex"])

In [13]:
df.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    1  22.0      1      0   7.2500        S
1         1       1    0  38.0      1      0  71.2833        C
2         1       3    0  26.0      0      0   7.9250        S
3         1       1    0  35.0      1      0  53.1000        S
4         0       3    1  35.0      0      0   8.0500        S

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 8 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    int64  
 3   Age       714 non-null    float64
 4   SibSp     891 non-null    int64  
 5   Parch     891 non-null    int64  
 6   Fare      891 non-null    float64
 7   Embarked  889 non-null    object 
dtypes: float64(2), int64(5), object(1)
memory usage: 55.8+ KB


In [15]:
df = df.fillna(df["Age"].mean().round())

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 8 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    int64  
 3   Age       891 non-null    float64
 4   SibSp     891 non-null    int64  
 5   Parch     891 non-null    int64  
 6   Fare      891 non-null    float64
 7   Embarked  891 non-null    object 
dtypes: float64(2), int64(5), object(1)
memory usage: 55.8+ KB


In [17]:
input_column = ["Pclass","Sex", "Age", "SibSp", "Parch", "Fare", "Embarked"]
output_column =['Survived']

In [18]:
X = df[input_column]
Y = df[output_column]


### Defining Entropy and Information Gain:

In [19]:
def entropy(col):
    counts_with_attributes = np.unique(col, return_counts = True)
    # This function will return the count of unique elements in the columns
    
    N = float(col.shape[0])
    # total no. of entries in that column
    
    print(counts_with_attributes)
    
    
    entropy = 0.0
    
# Now will iterate over the second part of count with attributes
    for ix in counts_with_attributes[1]: #ie array([2,5])
        p = ix/N
        entropy += (-1*p*np.log2(p))
    return entropy

In [20]:
x =np.unique([1,23,23,4,56,0], return_counts = True)
print(type(x))
print(len(x))

<class 'tuple'>
2


In [21]:
col = np.array([1, 1, 0, 0, 1, 0])
entropy(col) #np.unique returned that array is 0 and 1 type
# and it has 2 0's and five 1's.


(array([0, 1]), array([3, 3]))


1.0

In [22]:
X.columns

Index(['Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Embarked'], dtype='object')

In [38]:
#creating a function for spiltting the data
def divide_data(x_data, fkey, fval): 
    #work with pandas Data frames
    X_right = pd.DataFrame([], columns = X.columns)
    X_left  = pd.DataFrame([], columns = X.columns)

    for ix in range(X.shape[0]):
        val = X[fkey].loc[ix]
        # creating a threshold value for every row
        if val > fval:
            X_right = X_right.append(X.loc[ix]).va
        else:
            X_left  = X_left.append(X.loc[ix])
    return X_left, X_right
        

In [37]:
X_right = pd.DataFrame([], columns = X.columns)
X_right


Empty DataFrame
Columns: [Pclass, Sex, Age, SibSp, Parch, Fare, Embarked]
Index: []

In [29]:
X_left, X_right  = divide_data(X[:10], "Sex", 0.5)

AttributeError: 'DataFrame' object has no attribute 'append'

- We will pick which give maximum information gain
- afterwards I will divide my dataset into 2 parts
- for that we need threshold value as fvalue
- key value (fkey) will be attributes like income of a person may define he may  be able to buy the car or not.
- I can use mean to create a threshold value, male and female it will be 0.5.
- If sex value is greater than 0.5 he will be a male, else a female.
- We are going to use binary or more if required split at every node.    

In [34]:
def information_gain(x_data, fkey, fval):
    left, right = divide_data(x_data,fkey,fval)
    # % samples are on left hand side
    l = float(left.shape[0])/ x_data.shape[0]
    r = float(right.shape[0])/x_data.shape[0]

    # All examples come to one side!
    if left.shape[0] == 0 or right.shape[0] == 0: # means we cant get 
        # anything from the information gain
        return -1000000# Minimum information gain
    i_gain = entropy(x_data.Survived) -(l.entropy(left.survived) + r.entropy(right.Survived))
    #left side entropy and right side entropy
    return i_gain
    

In [35]:
# Test our functions:
from fx in X.columns:
    print(fx)
    print(information_gain(data_clean, fx, data_clean[fx].mean()))

SyntaxError: invalid syntax (3819162540.py, line 2)

In [43]:
class DecisionTree:
    #constructor
    def __init__(self, depth = 0 ,max_depth = 5):
        #decision tree can over fit
        self.left  = None
        self.right = None
        self.fkey  = None #key value , on which we wanna split like sex
        self.fval  = None # Threshold value = 0.5
        self.max_depth = max_depth #max depth = 5
        self.depth = depth #current depth = 3
        self.target = None 
# If a node has 50 examples, 40 yes and 10 are no at this
# leaf my entire prediction will be yes, because it has the
# majority class
    
# Building of trainning fuction
    def train(self, X_train):
        features = {"Pclass", "sex", "Age", "SibSp", "Parch", "Fare"}
        info_gains = []
        
#info gain for each feature:
        for ix in features:
            i_gain = information_gain(X_train, ix, X_train[ix].mean())
            info_gains.append(i_gain)

        self.fkey = features[np.argmax(info_gains)]
        self.fval = X_train[self.fkey].mean()
        

    #split data
        data_left, data_right = divide_data(X_train, self.fkey, self.fval)
        data_left = dat_left.reset_index(drop=True)# indexing again starts
    # with each node
        data_right = data_right.reset_index(drop=True)
### truly a left node -->> ist base case
        if data_left.shape[0] == 0 or dat_right.shape[0] ==0:
            if X_train.Survived.mean() >= 0.5:
             self.target = "Survive"
            else:
                self.target = "Dead"
            return 

### Stop early when depth >= max_depth

        if data_left.shape[0] == 0 or dat_right.shape[0] ==0:
            if X_train.Survived.mean() >= 0.5:
                 self.target = "Survive"
            else:
                self.target = "Dead"
            return 

# Recursive Case
        self.left = DecisionTree(depth=self.depth+1,max_depth=self.max_depth)
        self.left.train(data_left)

        self.right = DecisionTree(depth=self.depth +1, max_depth =self.max_depth)
        self.right.train(data_right)
        if X_train.Survived.mean() >= 0.5:
            self.target = "Survive"
        else:
            self.target = "Dead"
        return 
        
        
        

- If we build a tree which is like very deep
- 5 examples means 5 depth
- likely to overfitting
- I wont go deeper than that
- Poor generalization
- no. of nodes vs accuracy, Training accuracy increases against no. of nodes increases, but test accuracy or cross validation accuracy  will start decreasing after a  point.
- There are two :
- 1) Post Pount


In [44]:
d = DecisionTree()

In [45]:
d.train(df)

AttributeError: 'DataFrame' object has no attribute 'append'

### Train- Validation- Test Set Split

In [46]:
def predict(self, test):
    if test[self.fkey]>self.fval:
        #go to right
        if self.right in None:
            return self.target
        return self.right.predict(test)
    else:
        if self.left is None:
            return self.target
        return self.left.predict(test)    

### Train- Validation Test Ste Split

In [ ]:
split = int(0.7*df.shape[0])


### DT using SKLearn

In [47]:
from sklearn.tree import DecisionTreeClassifier

In [54]:
sk_tree = DecisionTreeClassifier?
# Gini index ,which  is mathematical function used foe entropy

Init signature:
DecisionTreeClassifier(
    *,
    criterion='gini',
    splitter='best',
    max_depth=None,
    min_samples_split=2,
    min_samples_leaf=1,
    min_weight_fraction_leaf=0.0,
    max_features=None,
    random_state=None,
    max_leaf_nodes=None,
    min_impurity_decrease=0.0,
    class_weight=None,
    ccp_alpha=0.0,
)
Docstring:     
A decision tree classifier.

Read more in the :ref:`User Guide <tree>`.

Parameters
----------
criterion : {"gini", "entropy", "log_loss"}, default="gini"
    The function to measure the quality of a split. Supported criteria are
    "gini" for the Gini impurity and "log_loss" and "entropy" both for the
    Shannon information gain, see :ref:`tree_mathematical_formulation`.

splitter : {"best", "random"}, default="best"
    The strategy used to choose the split at each node. Supported
    strategies are "best" to choose the best split and "random" to choose
    the best random split.

max_depth : int, default=None
    The maximum depth o

In [51]:
sk_tree = DecisionTreeClassifier(criterion='entropy')

In [ ]:
sk_tree.fit(train_data[input_column], train

In [4]:
import sys
!{sys.executable} -m pip install pydotplus

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 278.7/278.7 kB 9.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for pydotplus: filename=pydotplus-2.0.2-py3-none-any.whl size=24552 sha256=6ba249faf06d41a9ff5f2daf34cf265817d70ab42d5782585e6a01e159c72843
  Stored in directory: /Users/omkarthakur/Library/Caches/pip/wheels/bd/ce/e8/ff9d9c699514922f57caa22fbd55b0a32761114b4c4acc9e03
Successfully built pydotplus


In [7]:
pip install six

Note: you may need to restart the kernel to use updated packages.


### visualization of Decision tree

In [2]:
import pydotplus
sys.modules['sklearn.externals.six'] = six# data format
from Ipython.display import Image 
from sklearn.tree import export_graphviz

NameError: name 'six' is not defined

In [ ]:
dot_data = StringIO()
export_graphviz(sk_tree, out_file=dot, filled= True, rounded = True)


### Random Forest:

In [11]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score

In [14]:
rf = RandomForestClassifier(n_estimators = 10, criterion = "entropy")

In [15]:
rf.fit(X_train, Y_train)

NameError: name 'X_train' is not defined

In [16]:
rf.score(X_train, Y_train)

NameError: name 'X_train' is not defined

In [17]:
rf.score(X_test, Y_test)

NameError: name 'X_test' is not defined

In [ ]:
from sklearn.model_selection import cross_val_scorer
acc = cross_val_score(RandomForestClassifier(n_estimators=10, max_depth=5, criterion = "entropy"), X_train

In [ ]:
n_estimators is an hyper parat meter